Introducing improved Optional Checker

jyoo980 opened the pull request:
This PR introduces an improved version of the [Optional Checker](https://checkerframework.org/manual/#optional-checker) with increased precision around `Optional` values created from container (e.g., `java.util.stream.Stream`) types.

Merge with https://github.com/typetools/jdk/pull/222.
Supercedes https://github.com/typetools/checker-framework/pull/6685.

mernst reviewed (changes requested):
Thanks for this big feature.  I have a few comments and questions.

jyoo980 commented on checker/src/main/java/org/checkerframework/checker/optional/OptionalImplVisitor.java:
```diff
@@ -247,6 +279,8 @@ public void handleTernaryIsPresentGet(ConditionalExpressionTree tree) {
   private boolean sameExpression(ExpressionTree tree1, ExpressionTree tree2) {
```
I agree, but it seems like I would need to introduce a dependency on `org.checkerframework.dataflow.expression` if I made that refactoring; is this desirable?

mernst replied:
The `checker` package already depends on the `dataflow` package, so I think that is OK.

jyoo980 replied:
The `TreeUtils` class is not in the `checker` package. It is in the `javacutil` package, which I do not think declares a dependency on the `dataflow.expression` package, where `JavaExpression` is defined.

mernst replied:
@smillst Your thoughts?

smillst replied:
This method can't be in TreeUtils because the `javacutil` package should not depend on `dataflow`.  We could move JavaExpression to `javacutil`.  This would also required moving  method that use Nodes, i.e.`JavaExpression#fromNode(...)`, to somewhere else in the dataflow package.

I also don't see a better way to this comparison than what is here.

jyoo980 replied:
Thank you both for your input.

I've added additional context in the Javadoc for this method explaining why it isn't in `TreeUtils` in addition to a link to this discussion. I think this is adequate for now, but I am happy to revisit this if needed.

jyoo980 commented on dataflow/src/main/java/org/checkerframework/dataflow/expression/JavaExpression.java:
```diff
@@ -915,4 +917,89 @@ private static boolean isVarargsInvocation(
+   * Find the declaration of the receiver of a method call in a given method tree.
+   *
+   * <p>The receiver should appear in one of two places, either as a formal parameter to the method,
+   * or as a local variable.
```
To clarify, do you mean a case like below?
```java
import java.util.Optional;

import org.checkerframework.checker.optional.qual.Present;
import java.util.Optional;

@SuppressWarnings({
  "optional.field",
  "optional.parameter",
})
public class Test {

  private @Present Optional<String> opt = Optional.of("Hello");

  class Inner {

    private Optional<String> opt;

    void foo() {
      opt.get();
      Test.this.opt.get();
    }
  }
}
```

In this case, the Optional Checker correctly issues a warning at the line:
```java
opt.get();
```
But not at the line:
```java
Test.this.opt.get();
```

mernst replied:
Your example is a good test case, but it's not what I was thinking of.
I was thinking of:

```java
void m() {
  {
     Optional<String> opt;
     ...
  }
  ...
  {
     Optional<String> opt;
     ...
   }
}
```

Here there are two declarations of variables that have the same name but disjoint live ranges.

jyoo980 commented on dataflow/src/main/java/org/checkerframework/dataflow/expression/JavaExpression.java:
```diff
@@ -915,4 +917,89 @@ private static boolean isVarargsInvocation(
+        return param;
+      }
+    }
+    for (StatementTree statement : tree.getBody().getStatements()) {
```
I'm sorry to ask you to clarify, but could you give me an example of what you mean?

Do you mean something like the following?
```java
void foo() {

  {
    Object x = ...
  }
}
```
`x` is defined within a scope in `foo`, but it is not accessible outside of it. As a result, I think something like the code below would be impossible:

```java
void foo() {

  {
    Object x = ...
  }
  x.someMethod(); // Invalid Java: x is not accessible outside the scope
}
```

mernst replied:
The first example is what I had in mind.
You are right that the second code example is impossible.
But this is possible:
```java
void foo() {
  {
    Object x = ...;
    x.someMethod();
  }
}
```

jyoo980 replied:
I have added a [test case via this commit](https://github.com/typetools/checker-framework/pull/6901/commits/455adaeb4cba7c50d565eeca17634cea7f628f43)

jyoo980 commented on checker/src/main/java/org/checkerframework/checker/optional/OptionalImplVisitor.java:
```diff
@@ -285,21 +319,111 @@ public void handleConditionalStatementIsPresentGet(IfTree tree) {
     if (!(elseStmt == null
         || (elseStmt.getKind() == Tree.Kind.BLOCK
             && ((BlockTree) elseStmt).getStatements().isEmpty()))) {
       // else block is missing or is an empty block: "{}"
```
> The example in the Javadoc after "Pattern match for:" doesn't have an else clause.

I did not write this code (or that particular Javadoc example), but I think that is exactly why there's a `return`; this method does not match that case, and so no action should be taken. I may be misinterpreting the Javadoc or your comment, though.

jyoo980 commented on checker/src/main/java/org/checkerframework/checker/optional/OptionalImplVisitor.java:
```diff
@@ -159,7 +193,7 @@ private boolean isOptionalCreation(MethodInvocationTree methInvok) {
+   * @return true iff the method being called is Optional propagation: filter, flatMap, map, or
    */
   private boolean isOptionalPropagation(MethodInvocationTree methInvok) {
     ExecutableElement method = TreeUtils.elementFromUse(methInvok);
```
Thank you for the catch. I have updated the logic such that this warning/alternative refactoring is surfaced when the expression is pure [in this commit](https://github.com/typetools/checker-framework/pull/6901/commits/c651b2e45f167eb752d4923c7ebc6a41fab51934).

jyoo980 commented on checker/src/main/java/org/checkerframework/checker/optional/OptionalImplVisitor.java:
```diff
@@ -351,13 +581,11 @@ private void handleCompareToNull(BinaryTree tree) {
+    if (rightOp.getKind() == Tree.Kind.NULL_LITERAL && isOptionalType(TreeUtils.typeOf(leftOp))) {
       checker.reportWarning(tree, "optional.null.comparison");
     }
   }
```
I do not think the Javadoc for `OptionalImplVisitor#handleCreationElimination()` was correct.

The `introduce.eliminate` rule is not relevant to the preference for the `VAR.ifPresent(METHOD)` pattern (this assumes the `Optional` value already exists).

I have updated the Javadoc to instead explain each of `CREATION()`, `PROPAGATION()`, and `ELIMINATION()` and why this pattern is undesirable.

jyoo980 commented on checker/src/main/java/org/checkerframework/checker/optional/OptionalImplVisitor.java:
```diff
@@ -247,6 +279,8 @@ public void handleTernaryIsPresentGet(ConditionalExpressionTree tree) {
+    // Use transfer functions and Store entries.
     if (r1 != null && !r1.containsUnknown() && r2 != null && !r2.containsUnknown()) {
       return r1.equals(r2);
     } else {
```
Thanks for the catch, purity is now considered in the check for `handleConditionalStatementIsPresentGet` via [this commit](https://github.com/typetools/checker-framework/pull/6901/commits/6e587273596c3c68d1960f4420350c3e0aee37d9)

jyoo980 commented on checker/src/main/java/org/checkerframework/checker/optional/OptionalImplVisitor.java:
```diff
@@ -321,16 +447,120 @@ public void handleConditionalStatementIsPresentGet(IfTree tree) {
+        || isAnyFormalAnnotatedWithNonEmpty(methodDecl)) {
+      addMethodToVerifyWithNonEmptyChecker(methodDecl);
+    }
+    if (isReturnTypeAnnotatedWithNonEmpty(methodDecl)) {
+      namesOfMethodsToVerifyWithNonEmptyChecker.add(methodDecl.getName().toString());
```
I have documented the relationship between `namesOfMethodsToVerifyWithNonEmptyChecker` and `calleesToCallers` [via this commit](https://github.com/typetools/checker-framework/pull/6901/commits/497b8a755d2d32da6c3f01f4e4b93cad09a8c738).

mernst reviewed (approved):
Thanks!

smillst commented:
I think this needs a change log entry?  Something like whats in the PR description.  What do you think @mernst?  (Also, I have started reviewing this.)

mernst commented on docs/manual/non-empty-checker.tex:
```diff
@@ -11,6 +11,23 @@
 \<Deque.getFirst()>, \<Deque.removeFirst()>, \<Queue.remove()>, or
 \<Queue.element()>.
 
+Currently, the Non-Empty Checker operates as an auxiliary checker for the
+Optional Checker (\sectionpageref{optional-checker}).
```
I don't think this description is accurate any more.  I think the default modality is "normal" non-emptiness checking.

mernst commented:
> I think this needs a change log entry? Something like whats in the PR description.

That is reasonable.  I would express it as a precision improvement rather than a new version.
The behavior of the Non-Empty Checker has also changed, and that should also be documented in the changelog.

jyoo980 commented:
@mernst I have updated the changelog to account for changes in the Optional Checker.

> The behavior of the Non-Empty Checker has also changed, and that should also be documented in the changelog.

I'm not sure if this is true, the Non-Empty Checker was never publicly released and is currently not a user-facing checker (it is not listed in the manual).

smillst reviewed (changes requested):
I just have two comments.

jyoo980 commented on checker/src/main/java/org/checkerframework/checker/optional/OptionalChecker.java:
```diff
@@ -1,32 +1,37 @@
 
 /**
- * A type-checker that prevents misuse of the {@link java.util.Optional} class.
+ * A type-checker that prevents {@link java.util.NoSuchElementException} in the use of the {@link
+ * java.util.Optional} class.
+ *
+ * <p>The Optional Checker aggregates two checkers, the Non-Empty Checker and the
+ * OptionalImplChecker. The Optional Checker behaves as the ultimate parent of those two checkers
+ * and has no qualifiers or transfer function class of its own.
+ *
+ * <p>When the Non-Empty Checker is executed as a subchecker of the Optional Checker, it runs the
+ * OptionalImplChecker. It only type-checks the methods that the OptionalImplChecker passes to it.
+ *
+ * <p>The OptionalImplChecker implements a type system for Optional values (i.e., what users would
+ * expect the Optional Checker to implement). It reads programmer-written annotations from the
+ * Non-Empty type system, and keeps track of the methods that the Non-Empty Checker should check.
```
@smillst I have written a summary of the interactions between the `OptionalChecker`, `NonEmptyChecker`, and the `OptionalImplChecker` here

smillst replied:
This great thanks!  It should just be a `\\` comment since it contains implementation details.

jyoo980 commented:
Fixed!

– james

> On Jan 8, 2025, at 11:03 AM, Suzanne Millstein ***@***.***> wrote:
> 
> 
> @smillst commented on this pull request.
> 
> In checker/src/main/java/org/checkerframework/checker/optional/OptionalChecker.java <https://github.com/typetools/checker-framework/pull/6901#discussion_r1907670574>:
> 
> > + * A type-checker that prevents ***@***.*** java.util.NoSuchElementException} in the use of the ***@***.***
> + * java.util.Optional} class.
> + *
> + * <p>The Optional Checker aggregates two checkers, the Non-Empty Checker and the
> + * OptionalImplChecker. The Optional Checker behaves as the ultimate parent of those two checkers
> + * and has no qualifiers or transfer function class of its own.
> + *
> + * <p>When the Non-Empty Checker is executed as a subchecker of the Optional Checker, it runs the
> + * OptionalImplChecker. It only type-checks the methods that the OptionalImplChecker passes to it.
> + *
> + * <p>The OptionalImplChecker implements a type system for Optional values (i.e., what users would
> + * expect the Optional Checker to implement). It reads programmer-written annotations from the
> + * Non-Empty type system, and keeps track of the methods that the Non-Empty Checker should check.
> This great thanks! It should just be a \\ comment since it contains implementation details.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/typetools/checker-framework/pull/6901#discussion_r1907670574>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFZ3EEDBCSJUUEZILTGUFY32JVZB3AVCNFSM6AAAAABSLQHRAOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKMZXHE3TIOJQGU>.
> You are receiving this because you were assigned.
>

smillst reviewed (approved):
Looks good to me.  Thanks!
